app/controllers/courses/subjects_controller.rb

```diff
@@ -6,13 +6,16 @@ class SubjectsController < ApplicationController
     include CourseBasicDetailConcern
 
     def update
-      master_subject_id = params.dig(:course, :master_subject_id)
-      master_subject_hash = @course.meta[:edit_options][:subjects].find do |subject|
-        subject[:id] == master_subject_id
-      end
-      master_subject = Subject.new(master_subject_hash.to_h)
-
-      if @course.update(subjects: [master_subject])
+      if subjects_match?(new_subjects, old_non_language_subjects)
```

Comment:
Wonder if there may be an opportunity to increase readability of this if else and move the logic out into private methods?
Pls forgive me for the bad pseudo-code
e.g.

```
return [ if there are errors]
if no_changes_to_subjects?
   [Do the thing]
else
   [update the thing]
end
```

Reply:
I think the problem here is that the `update()` both returns errors and does work so we can't really know whether or not there are errors beforehand. I don't know that there's a way to get around that without introducing a race condition.